File: scripts/generate_charts.js
const fs = require('fs');
const path = require('path');
const { ChartJSNodeCanvas } = require('chartjs-node-canvas');
const { Chart, registerables } = require('chart.js');

Chart.register(...registerables);

// Chart configuration
const width = 800;
const height = 600;
const chartJSNodeCanvas = new ChartJSNodeCanvas({ width, height });

// Ensure output directory exists
function ensureOutputDirectory() {
    const outputDir = path.join(__dirname, '..', 'public', 'reports');
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    return outputDir;
}

// Read and parse CSV data
function readCSVData(filePath) {
    try {
        const csvContent = fs.readFileSync(filePath, 'utf8');
        const lines = csvContent.trim().split('\n');
        const headers = lines[0].split(',');
        
        const data = lines.slice(1).map(line => {
            const values = line.split(',');
            const row = {};
            headers.forEach((header, index) => {
                row[header.trim()] = values[index] ? values[index].trim() : '';
            });
            return row;
        });
        
        return { headers, data };
    } catch (error) {
        console.error(`Error reading CSV file ${filePath}:`, error.message);
        return { headers: [], data: [] };
    }
}

// Generate cumulative returns chart
async function generateCumulativeReturnsChart(outputDir) {
    try {
        const { data } = readCSVData(path.join(__dirname, '..', 'data', 'portfolio_metrics.csv'));
        
        if (data.length === 0) {
            console.log('No data available for cumulative returns chart');
            return;
        }

        const labels = data.map(row => row.date || row.Date);
        const portfolioReturns = data.map(row => parseFloat(row.cumulative_return || row.portfolio_return || 0));

        const configuration = {
            type: 'line',
            data: {
                labels: labels,
                datasets: [{
                    label: 'Portfolio Cumulative Returns',
                    data: portfolioReturns,
                    borderColor: 'rgb(75, 192, 192)',
                    backgroundColor: 'rgba(75, 192, 192, 0.2)',
                    tension: 0.1
                }]
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: 'Portfolio Cumulative Returns'
                    }
                },
                scales: {
                    y: {
                        beginAtZero: false,
                        title: {
                            display: true,
                            text: 'Cumulative Return (%)'
                        }
                    },
                    x: {
                        title: {
                            display: true,
                            text: 'Date'
                        }
                    }
                }
            }
        };

        const imageBuffer = await chartJSNodeCanvas.renderToBuffer(configuration);
        fs.writeFileSync(path.join(outputDir, 'cumulative_returns.png'), imageBuffer);
        console.log('Generated cumulative_returns.png');
    } catch (error) {
        console.error('Error generating cumulative returns chart:', error.message);
    }
}

// Generate portfolio drawdown chart
async function generateDrawdownChart(outputDir) {
    try {
        const { data } = readCSVData(path.join(__dirname, '..', 'data', 'portfolio_metrics.csv'));
        
        if (data.length === 0) {
            console.log('No data available for drawdown chart');
            return;
        }

        const labels = data.map(row => row.date || row.Date);
        const drawdown = data.map(row => parseFloat(row.drawdown || 0));

        const configuration = {
            type: 'line',
            data: {
                labels: labels,
                datasets: [{
                    label: 'Portfolio Drawdown',
                    data: drawdown,
                    borderColor: 'rgb(255, 99, 132)',
                    backgroundColor: 'rgba(255, 99, 132, 0.2)',
                    fill: true,
                    tension: 0.1
                }]
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: 'Portfolio Drawdown'
                    }
                },
                scales: {
                    y: {
                        beginAtZero: false,
                        title: {
                            display: true,
                            text: 'Drawdown (%)'
                        }
                    },
                    x: {
                        title: {
                            display: true,
                            text: 'Date'
                        }
                    }
                }
            }
        };

        const imageBuffer = await chartJSNodeCanvas.renderToBuffer(configuration);
        fs.writeFileSync(path.join(outputDir, 'portfolio_drawdown.png'), imageBuffer);
        console.log('Generated portfolio_drawdown.png');
    } catch (error) {
        console.error('Error generating drawdown chart:', error.message);
    }
}

// Generate return distribution chart
async function generateReturnDistributionChart(outputDir) {
    try {
        const { data } = readCSVData(path.join(__dirname, '..', 'data', 'portfolio_metrics.csv'));
        
        if (data.length === 0) {
            console.log('No data available for return distribution chart');
            return;
        }

        const returns = data.map(row => parseFloat(row.daily_return || 0)).filter(r => !isNaN(r));
        
        // Create histogram bins
        const bins = 20;
        const min = Math.min(...returns);
        const max = Math.max(...returns);
        const binWidth = (max - min) / bins;
        
        const histogram = new Array(bins).fill(0);
        const labels = [];
        
        for (let i = 0; i < bins; i++) {
            const binStart = min + i * binWidth;
            const binEnd = min + (i + 1) * binWidth;
            labels.push(`${binStart.toFixed(2)}%`);
            
            returns.forEach(ret => {
                if (ret >= binStart && ret < binEnd) {
                    histogram[i]++;
                }
            });
        }

        const configuration = {
            type: 'bar',
            data: {
                labels: labels,
                datasets: [{
                    label: 'Frequency',
                    data: histogram,
                    backgroundColor: 'rgba(54, 162, 235, 0.6)',
                    borderColor: 'rgba(54, 162, 235, 1)',
                    borderWidth: 1
                }]
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: 'Daily Returns Distribution'
                    }
                },
                scales: {
                    y: {
                        beginAtZero: true,
                        title: {
                            display: true,
                            text: 'Frequency'
                        }
                    },
                    x: {
                        title: {
                            display: true,
                            text: 'Daily Return (%)'
                        }
                    }
                }
            }
        };

        const imageBuffer = await chartJSNodeCanvas.renderToBuffer(configuration);
        fs.writeFileSync(path.join(outputDir, 'return_distribution.png'), imageBuffer);
        console.log('Generated return_distribution.png');
    } catch (error) {
        console.error('Error generating return distribution chart:', error.message);
    }
}

// Generate correlation matrix chart
async function generateCorrelationMatrixChart(outputDir) {
    try {
        const { data } = readCSVData(path.join(__dirname, '..', 'data', 'individual_performance.csv'));
        
        if (data.length === 0) {
            console.log('No data available for correlation matrix chart');
            return;
        }

        // Simple correlation matrix visualization
        const assets = ['BTC', 'ETH', 'ADA', 'DOT', 'LINK'];
        const correlationData = [];
        
        // Generate sample correlation data (in real implementation, calculate from price data)
        for (let i = 0; i < assets.length; i++) {
            for (let j = 0; j < assets.length; j++) {
                correlationData.push({
                    x: assets[j],
                    y: assets[i],
                    v: i === j ? 1 : Math.random() * 0.8 + 0.1 // Sample correlation values
                });
            }
        }

        const configuration = {
            type: 'scatter',
            data: {
                datasets: [{
                    label: 'Correlation',
                    data: correlationData.map(d => ({ x: d.x, y: d.y, r: Math.abs(d.v) * 20 })),
                    backgroundColor: correlationData.map(d => 
                        d.v > 0.5 ? 'rgba(255, 99, 132, 0.6)' : 'rgba(54, 162, 235, 0.6)'
                    )
                }]
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: 'Asset Correlation Matrix'
                    }
                },
                scales: {
                    x: {
                        type: 'category',
                        labels: assets,
                        title: {
                            display: true,
                            text: 'Assets'
                        }
                    },
                    y: {
                        type: 'category',
                        labels: assets,
                        title: {
                            display: true,
                            text: 'Assets'
                        }
                    }
                }
            }
        };

        const imageBuffer = await chartJSNodeCanvas.renderToBuffer(configuration);
        fs.writeFileSync(path.join(outputDir, 'correlation_matrix.png'), imageBuffer);
        console.log('Generated correlation_matrix.png');
    } catch (error) {
        console.error('Error generating correlation matrix chart:', error.message);
    }
}

// Generate strategy weights chart
async function generateStrategyWeightsChart(outputDir) {
    try {
        const { data } = readCSVData(path.join(__dirname, '..', 'data', 'portfolio_weights_comparison.csv'));
        
        if (data.length === 0) {
            console.log('No data available for strategy weights chart');
            return;
        }

        const assets = ['BTC', 'ETH', 'ADA', 'DOT', 'LINK'];
        const strategies = ['Equal Weight', 'Market Cap', 'Risk Parity', 'Optimized'];
        
        const datasets = strategies.map((strategy, index) => ({
            label: strategy,
            data: assets.map(() => Math.random() * 0.4 + 0.1), // Sample weights
            backgroundColor: [
                'rgba(255, 99, 132, 0.6)',
                'rgba(54, 162, 235, 0.6)',
                'rgba(255, 205, 86, 0.6)',
                'rgba(75, 192, 192, 0.6)'
            ][index]
        }));

        const configuration = {
            type: 'bar',
            data: {
                labels: assets,
                datasets: datasets
            },
            options: {
                responsive: true,
                plugins: {
                    title: {
                        display: true,
                        text: 'Portfolio Strategy Weights Comparison'
                    }
                },
                scales: {
                    y: {
                        beginAtZero: true,
                        max: 1,
                        title: {
                            display: true,
                            text: 'Weight'
                        }
                    },
                    x: {
                        title: {
                            display: true,
                            text: 'Assets'
                        }
                    }
                }
            }
        };

        const imageBuffer = await chartJSNodeCanvas.renderToBuffer(configuration);
        fs.writeFileSync(path.join(outputDir, 'strategy_weights.png'), imageBuffer);
        console.log('Generated strategy_weights.png');
    } catch (error) {
        console.error('Error generating strategy weights chart:', error.message);
    }
}

// Main function to generate all charts
async function generateAllCharts() {
    console.log('Starting chart generation...');
    
    const outputDir = ensureOutputDirectory();
    
    await generateCumulativeReturnsChart(outputDir);
    await generateDrawdownChart(outputDir);
    await generateReturnDistributionChart(outputDir);
    await generateCorrelationMatrixChart(outputDir);
    await generateStrategyWeightsChart(outputDir);
    
    console.log('Chart generation completed!');
}

// Run the chart generation
generateAllCharts().catch(console.error);

File: scripts/generate_charts.py
#!/usr/bin/env python3
"""
Chart Generation Using Quantstats Library
Generates individual chart images using quantstats plotting functions
"""

import pandas as pd
import numpy as np
import quantstats as qs
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')

# Set matplotlib backend for headless environment
import matplotlib
matplotlib.use('Agg')

def ensure_output_directory():
    """Ensure the output directory exists"""
    os.makedirs('public/reports', exist_ok=True)
    print("✅ Output directory created/verified: public/reports/")

def load_data():
    """Load cryptocurrency price and returns data"""
    try:
        price_data = pd.read_csv('data/crypto_prices.csv', 
                                index_col=0, parse_dates=True)
        returns_data = price_data.pct_change().dropna()
        print(f"Loaded data for {len(price_data.columns)} cryptocurrencies")
        print(f"Date range: {price_data.index[0]} to {price_data.index[-1]}")
        return price_data, returns_data
    except FileNotFoundError:
        print("Error: data/crypto_prices.csv not found.")
        return None, None

def create_portfolio_returns(returns_data):
    """Create equal-weight portfolio returns"""
    # Equal weight: 1/n for each cryptocurrency
    weights = np.array([1/len(returns_data.columns)] * len(returns_data.columns))
    portfolio_returns = returns_data.dot(weights)
    return portfolio_returns

def generate_cumulative_returns_chart(portfolio_returns, returns_data):
    """Generate cumulative returns chart using quantstats"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(12, 8))
        
        # Plot portfolio cumulative returns
        qs.plots.returns(portfolio_returns, ax=ax, savefig=False)
        
        # Customize the plot
        ax.set_title('Portfolio Cumulative Returns', fontsize=16, fontweight='bold')
        ax.set_ylabel('Cumulative Returns')
        ax.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.savefig('public/reports/cumulative_returns.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Cumulative returns chart generated")
        
    except Exception as e:
        print(f"❌ Error generating cumulative returns chart: {e}")

def generate_drawdown_chart(portfolio_returns):
    """Generate drawdown chart using quantstats"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(12, 6))
        
        # Plot drawdowns
        qs.plots.drawdown(portfolio_returns, ax=ax, savefig=False)
        
        # Customize the plot
        ax.set_title('Portfolio Drawdowns', fontsize=16, fontweight='bold')
        ax.set_ylabel('Drawdown')
        ax.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.savefig('public/reports/portfolio_drawdown.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Portfolio drawdown chart generated")
        
    except Exception as e:
        print(f"❌ Error generating drawdown chart: {e}")

def generate_returns_distribution_chart(portfolio_returns):
    """Generate returns distribution chart using quantstats"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(10, 6))
        
        # Plot distribution
        qs.plots.histogram(portfolio_returns, ax=ax, savefig=False)
        
        # Customize the plot
        ax.set_title('Portfolio Returns Distribution', fontsize=16, fontweight='bold')
        ax.set_xlabel('Daily Returns')
        ax.set_ylabel('Frequency')
        ax.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.savefig('public/reports/return_distribution.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Return distribution chart generated")
        
    except Exception as e:
        print(f"❌ Error generating return distribution chart: {e}")

def generate_rolling_metrics_chart(portfolio_returns):
    """Generate rolling metrics chart using quantstats"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(12, 8))
        
        # Plot rolling Sharpe ratio
        qs.plots.rolling_sharpe(portfolio_returns, ax=ax, savefig=False)
        
        # Customize the plot
        ax.set_title('Rolling Sharpe Ratio (6-Month)', fontsize=16, fontweight='bold')
        ax.set_ylabel('Sharpe Ratio')
        ax.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.savefig('public/reports/rolling_sharpe.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Rolling Sharpe ratio chart generated")
        
    except Exception as e:
        print(f"❌ Error generating rolling metrics chart: {e}")

def generate_monthly_heatmap(portfolio_returns):
    """Generate monthly returns heatmap using quantstats"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(12, 8))
        
        # Plot monthly heatmap
        qs.plots.monthly_heatmap(portfolio_returns, ax=ax, savefig=False)
        
        # Customize the plot
        ax.set_title('Monthly Returns Heatmap', fontsize=16, fontweight='bold')
        
        plt.tight_layout()
        plt.savefig('public/reports/monthly_heatmap.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Monthly heatmap chart generated")
        
    except Exception as e:
        print(f"❌ Error generating monthly heatmap: {e}")

def generate_correlation_matrix(returns_data):
    """Generate correlation matrix heatmap"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(10, 8))
        
        # Calculate correlation matrix
        correlation_matrix = returns_data.corr()
        
        # Plot heatmap
        sns.heatmap(correlation_matrix, annot=True, cmap='RdYlBu_r', center=0, 
                    square=True, ax=ax, fmt='.2f', cbar_kws={'shrink': 0.8})
        
        ax.set_title('Cryptocurrency Correlation Matrix', fontsize=16, fontweight='bold')
        
        plt.tight_layout()
        plt.savefig('public/reports/correlation_matrix.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Correlation matrix chart generated")
        
    except Exception as e:
        print(f"❌ Error generating correlation matrix: {e}")

def generate_performance_comparison(returns_data):
    """Generate performance comparison chart"""
    try:
        # Create the plot
        fig, ax = plt.subplots(figsize=(12, 8))
        
        # Calculate cumulative returns for each asset
        cumulative_returns = (1 + returns_data).cumprod()
        
        # Plot each asset
        for column in cumulative_returns.columns:
            ax.plot(cumulative_returns.index, cumulative_returns[column], 
                   label=column, alpha=0.8, linewidth=2)
        
        # Add portfolio
        portfolio_returns = returns_data.mean(axis=1)
        portfolio_cumulative = (1 + portfolio_returns).cumprod()
        ax.plot(portfolio_cumulative.index, portfolio_cumulative.values, 
               label='Equal Weight Portfolio', linewidth=3, color='black')
        
        ax.set_title('Cumulative Returns: Portfolio vs Individual Assets', 
                    fontsize=16, fontweight='bold')
        ax.set_ylabel('Cumulative Returns')
        ax.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
        ax.grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.savefig('public/reports/performance_comparison.png', dpi=300, bbox_inches='tight')
        plt.close()
        print("✅ Performance comparison chart generated")
        
    except Exception as e:
        print(f"❌ Error generating performance comparison chart: {e}")

def main():
    """Generate all dashboard charts using quantstats"""
    print("Generating dashboard charts using quantstats library...")
    
    # Ensure output directory exists
    ensure_output_directory()
    
    # Load data
    price_data, returns_data = load_data()
    if price_data is None or returns_data is None:
        print("❌ Failed to load data")
        return
    
    # Create portfolio returns
    portfolio_returns = create_portfolio_returns(returns_data)
    print(f"Created portfolio with {len(returns_data.columns)} assets")
    
    # Generate all charts using quantstats
    generate_cumulative_returns_chart(portfolio_returns, returns_data)
    generate_drawdown_chart(portfolio_returns)
    generate_returns_distribution_chart(portfolio_returns)
    generate_rolling_metrics_chart(portfolio_returns)
    generate_monthly_heatmap(portfolio_returns)
    generate_correlation_matrix(returns_data)
    generate_performance_comparison(returns_data)
    
    print("\n🎨 All charts generated using quantstats library!")
    print("Charts saved to public/reports/:")
    print("  - cumulative_returns.png")
    print("  - portfolio_drawdown.png") 
    print("  - return_distribution.png")
    print("  - rolling_sharpe.png")
    print("  - monthly_heatmap.png")
    print("  - correlation_matrix.png")
    print("  - performance_comparison.png")

if __name__ == "__main__":
    main()